### ISSUE TITLE
version: 5.15.7 set context SDImageCoderDecodeThumbnailPixelSize  memory cache will disabled. 

### ISSUE THREAD

**Issue by LeeZhe** [SDWebImage/SDWebImage#3656]
### New Issue Checklist

* [x] I have read and understood the [CONTRIBUTING guide](https://github.com/rs/SDWebImage/blob/master/.github/CONTRIBUTING.md)
* [x] I have read the [Documentation](http://cocoadocs.org/docsets/SDWebImage/)
* [x] I have searched for a similar issue in the [project](https://github.com/rs/SDWebImage/issues) and found none

### Issue Info

 Info                    | Value                               |
-------------------------|-------------------------------------|
 Platform Name           | e.g. ios / macos / tvos / watchos
 Platform Version        | e.g. 16.2 / 10.14.0 / 12.0 / 5.0
 SDWebImage Version      | e.g. 5.15.7 / 4.4.0
 Integration Method      | e.g. carthage / cocoapods / manually
 Xcode Version           | e.g. Xcode 15.1 / Xcode 15.1
 Repro rate              | e.g. all the time (100%) / sometimes x% / only once
 Repro with our demo prj | e.g. does it happen with our demo project?
 Demo project link       | e.g. link to a demo project that highlights the issue

### Issue Description and Steps

Please fill in the detailed description of the issue (full output of any stack trace, compiler error, ...) and the steps to reproduce the issue.

the code is 
```
SDWebImageManager.m
 BOOL isThumbnail = originalImage.sd_isThumbnail;
    NSData *cacheData = originalData;
    UIImage *cacheImage = originalImage;
    if (isThumbnail) {
        cacheData = nil; // thumbnail don't store full size data
        originalImage = nil; // thumbnail don't have full size image
    }
```
```
SDImageCache.m
            CGSize thumbnailSize = CGSizeZero;
            NSValue *thumbnailSizeValue = context[SDWebImageContextImageThumbnailPixelSize];
            if (thumbnailSizeValue != nil) {
        #if SD_MAC
                thumbnailSize = thumbnailSizeValue.sizeValue;
        #else
                thumbnailSize = thumbnailSizeValue.CGSizeValue;
        #endif
            }
            if (thumbnailSize.width > 0 && thumbnailSize.height > 0) {
                // Query full size cache key which generate a thumbnail, should not write back to full size memory cache
                shouldCacheToMomery = NO;
            }
```

the image cache is always from disk， why not store to memory cache?  if store memory What impacts will it have？

**Comment by dreampiggy** [comment]
I guess you misunderstand that code.

That code is used to re-encode data from thumbnail image

**Comment by dreampiggy** [comment]
We have control about the cache behavior for thumbnail.

Check SDWebImageContextOriginalStoreCacheType

**Comment by LeeZhe** [comment]
@dreampiggy  In order for the encoded image to occupy less memory, when passing SDWebImageContextImageThumbnailPixelSize to the context, the code will determine that shouldCacheToMemory is false. At this time, when loading the same image, the cacheType of subsequent images is disk. When loading multiple images simultaneously from disk, the disk io queue takes up time and the interface will flicker. How can I pass in SDWebImageContextImageThumbnailPixelSize to allow images to also enjoy memory caching?

**Comment by LeeZhe** [comment]
When I comment on the following code 

`SDWebImageManager.m
 BOOL isThumbnail = originalImage.sd_isThumbnail;
    NSData *cacheData = originalData;
    UIImage *cacheImage = originalImage;
    if (isThumbnail) {
        cacheData = nil; // thumbnail don't store full size data
        originalImage = nil; // thumbnail don't have full size image
    }
SDImageCache.m
            CGSize thumbnailSize = CGSizeZero;
            NSValue *thumbnailSizeValue = context[SDWebImageContextImageThumbnailPixelSize];
            if (thumbnailSizeValue != nil) {
        #if SD_MAC
                thumbnailSize = thumbnailSizeValue.sizeValue;
        #else
                thumbnailSize = thumbnailSizeValue.CGSizeValue;
        #endif
            }
            if (thumbnailSize.width > 0 && thumbnailSize.height > 0) {
                // Query full size cache key which generate a thumbnail, should not write back to full size memory cache
                shouldCacheToMomery = NO;
            }
`

and try to pass in SDWebImageContextImageThumbnailPixelSize, I can now enjoy memory caching. I would like to inquire about the other effects of commenting on the above code?

**Comment by LeeZhe** [comment]
When I comment on the following code

```

SDWebImageManager.m
 BOOL isThumbnail = originalImage.sd_isThumbnail;
    NSData *cacheData = originalData;
    UIImage *cacheImage = originalImage;
    if (isThumbnail) {
        cacheData = nil; // thumbnail don't store full size data
        originalImage = nil; // thumbnail don't have full size image
    }
SDImageCache.m
            CGSize thumbnailSize = CGSizeZero;
            NSValue *thumbnailSizeValue = context[SDWebImageContextImageThumbnailPixelSize];
            if (thumbnailSizeValue != nil) {
        #if SD_MAC
                thumbnailSize = thumbnailSizeValue.sizeValue;
        #else
                thumbnailSize = thumbnailSizeValue.CGSizeValue;
        #endif
            }
            if (thumbnailSize.width > 0 && thumbnailSize.height > 0) {
                // Query full size cache key which generate a thumbnail, should not write back to full size memory cache
                shouldCacheToMomery = NO;
            }

```

and try to pass in SDWebImageContextImageThumbnailPixelSize, I can now enjoy memory caching. I would like to inquire about the other effects of commenting on the above code?

**Comment by dreampiggy** [comment]
Seems I can not can what's your feature request or desired behavior. It seems you're talking about a bug behavior but I'm sure it's already fixed in history version.

Check our Demo, which indeed use thumbail, it will query from both memory cache and disk cache.

Or, provide a demo for your request ?

**Comment by dreampiggy** [comment]
You can also send me more information via my email or social link.

**Comment by dreampiggy** [comment]
Are you want the following behavior ?

1. store full NSData into disk
2. stoee thumbnail NSData into disk
3. store thumnail UIImage into disk
4. store full UIImage into memory


For this behavior, check that two enum, the storeCahceTyoe and originalStoreCacheType, if you set both to `all`, then 1-4 will both becomes true.

**Comment by LeeZhe** [comment]
我中文直接跟你沟通吧，demo已经发给你了。你可以下载demo查看。
我的问题是：

// 问题：当我传入imageThumbnailPixelSize 时，给到的cacheType都是来自disk。也就是说内存缓存失效了。
                        // 原因: 由于下面的代码
                        /*
                         UIImage *diskImage;
                         if (image) {
                             // the image is from in-memory cache, but need image data
                             diskImage = image;
                         } else if (diskData) {
                             BOOL shouldCacheToMomery = YES;
                             if (context[SDWebImageContextStoreCacheType]) {
                                 SDImageCacheType cacheType = [context[SDWebImageContextStoreCacheType] integerValue];
                                 shouldCacheToMomery = (cacheType == SDImageCacheTypeAll || cacheType == SDImageCacheTypeMemory);
                             }
                             CGSize thumbnailSize = CGSizeZero;
                             NSValue *thumbnailSizeValue = context[SDWebImageContextImageThumbnailPixelSize];
                             if (thumbnailSizeValue != nil) {
                         #if SD_MAC
                                 thumbnailSize = thumbnailSizeValue.sizeValue;
                         #else
                                 thumbnailSize = thumbnailSizeValue.CGSizeValue;
                         #endif
                             }
                             if (thumbnailSize.width > 0 && thumbnailSize.height > 0) {
                                 // Query full size cache key which generate a thumbnail, should not write back to full size memory cache
                                 shouldCacheToMomery = NO;
                             }
                             // Special case: If user query image in list for the same URL, to avoid decode and write **same** image object into disk cache multiple times, we query and check memory cache here again.
                             if (shouldCacheToMomery && self.config.shouldCacheImagesInMemory) {
                                 diskImage = [self.memoryCache objectForKey:key];
                             }
                             // decode image data only if in-memory cache missed
                             if (!diskImage) {
                                 diskImage = [self diskImageForKey:key data:diskData options:options context:context];
                                 if (shouldCacheToMomery && diskImage && self.config.shouldCacheImagesInMemory) {
                                     NSUInteger cost = diskImage.sd_memoryCost;
                                     [self.memoryCache setObject:diskImage forKey:key cost:cost];
                                 }
                             }
                         **/
                        //疑惑：当检测到上层传入 SDWebImageContextOption.imageThumbnailPixelSize 有值时，就不会检索内存缓存，也不会将解码后的图片存储到内存缓存。
                        //业务加载表现: 当内存缓存都失效时，此时就会出现disk队列等待，多次刷新页面时，加载同一张图，由于缓存有队列控制，加载时间会拉长，用户就会感觉到出现图片闪烁的情况。
                        // 暂时解决方案: 当我把上面相关的代码注释，在检测到`SDWebImageContextOption.imageThumbnailPixelSize` 有值时也写入image到内存缓存，就不会出现闪烁的情况。
                        // 我想问的是:
                        // 1. 如果我设置了 SDWebImageContextOption.imageThumbnailPixelSize，如何能使得内存缓存生效。
                        // 2. 我把 shouldCacheToMomery 为NO相关代码注释后，image 写入内存或者，读取也读取内存缓存，会不会有其他的问题。

**Comment by LeeZhe** [comment]
我简单了个demo ，zip文件已经通过你的邮箱：[lizhuoli1126@126.com](mailto:lizhuoli1126@126.com)发你了。你可以跑起来看看。

**Comment by dreampiggy** [comment]
> 当检测到上层传入 SDWebImageContextOption.imageThumbnailPixelSize 有值时，就不会检索内存缓存，也不会将解码后的图片存储到内存缓存

不符合预期，至少不符合我当时设计这个Feature的初衷。

1. 先检查thumbnail size是否由缓存
1. 2. 如果没有缓存，检查full size
1. 2. 如果存在，重新decode生成thumbnail
2. 下载full data
3. decode得到thumbnail
4. thumbnail写入memory cache（cache key为thumbnail的）
5. full data写入disk cache（cache key为full的）

设计的Feature预期流程如上

**Comment by LeeZhe** [comment]
> > 当检测到上层传入 SDWebImageContextOption.imageThumbnailPixelSize 有值时，就不会检索内存缓存，也不会将解码后的图片存储到内存缓存
> 
> 不符合预期，至少不符合我当时设计这个Feature的初衷。
> 
> 1. 先检查thumbnail size是否缓存
> 2. 2. 如果没有缓存，检查full size
> 3. 2. 如果full size存在，重新decode生成thumbnail，返回
> 4. 下载full data
> 5. decode得到thumbnail（注意内存中从来没有full size的image出现）
> 6. thumbnail写入memory cache（cache key为thumbnail的）
> 7. full data写入disk cache（cache key为full的）
> 
> 设计的Feature预期流程如上（全默认值情况下，你自己传入其他context控制或者改了SDImageCacheConfig除外）

我这边尝试把  下面的代码注释，就能享受内存缓存了。

```
CGSize thumbnailSize = CGSizeZero;
NSValue *thumbnailSizeValue = context[SDWebImageContextImageThumbnailPixelSize];
if (thumbnailSizeValue != nil) {
#if SD_MAC
thumbnailSize = thumbnailSizeValue.sizeValue;
#else
thumbnailSize = thumbnailSizeValue.CGSizeValue;
#endif
}
if (thumbnailSize.width > 0 && thumbnailSize.height > 0) {
// Query full size cache key which generate a thumbnail, should not write back to full size memory cache
shouldCacheToMomery = NO;
}
```
**假设我不改SD的源码。**我怎么能够传入imageThumbnailPixelSize，够享受加载同一个url时，thumbnail能从内存中返回，已提高接在速度，已解决刷新页面的闪烁问题呢?

**Comment by dreampiggy** [comment]
简单理了下，你的测试Demo有一个假设是错误的：

你假设了磁盘读写是“同步的”，即，

`SDWebImagePrefetcher.shared.prefetchURLs`

的callback时刻，磁盘已经写入完成，这个并不是正确的。你需要传入额外的options

`SDWebImageWaitStoreCache`

**Comment by dreampiggy** [comment]
你读一下changelog，这里专门提到了这个背景：

### Behavior Changes
- Introduce `SDWebImageContextImageDecodeOptions`, deprecate `SDImageCoderWebImageContext` #3421
- - Note the deprecated `SDImageCoderWebImageContext` is no longer used. Update your codebase to use `SDWebImageContextImageDecodeOptions` to pass extra information from manager to coder. Loaded image will also store the same information in `image.sd_decodeOptions`
- Refactor the hack when multiple different thumbnail image requests for same url called at the same time #3423
- - Now the SDWebImageDownloader can callback each thumbnail request for same url with the correct image size as well (previously can't)
- - Now if manager's callback image is thumbnail (`image.sd_isThumbnail == YES`), or transformed (`image.sd_isTransformed == YES`), the callback data will be nil. This because the image does not match downloaded data actually. If you really need original full size data, query the disk cache again using the original key. You may need `SDWebImageWaitStoreCache` options as well. (previously behavior is undefined, may callback with full size data, or encoded small size data, or even nil)

**Comment by dreampiggy** [comment]
![image](https://github.com/SDWebImage/SDWebImage/assets/6919743/d3e00ef9-10ef-495f-b778-b61a98e1b3c4)

什么意思呢，在你假设的“预加载完成“的时刻，Disk Cache是空的，写入还没完成，正在异步写

**Comment by LeeZhe** [comment]
> ![image](https://private-user-images.githubusercontent.com/6919743/294832487-d3e00ef9-10ef-495f-b778-b61a98e1b3c4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQ2OTg2NzksIm5iZiI6MTcwNDY5ODM3OSwicGF0aCI6Ii82OTE5NzQzLzI5NDgzMjQ4Ny1kM2UwMGVmOS0xMGVmLTQ5NWYtYjc3OC1iNjFhOThlMWIzYzQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDEwOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAxMDhUMDcxOTM5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NTM0YTAxMzEzMzE2ZWY1NDA5NjZkZjAzOTAxNmVkODQ0OGNjODc2M2NkNWFiZGVhZTRjMmFiZjRmNDBmNzliYyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.-le9dpHg24rTVJFWV18hn0wWkGA_asJxhyxjalnGi7g)
> 
> 什么意思呢，在你假设的“预加载完成“的时刻，Disk Cache是空的，写入还没完成，正在异步写
> 
> 因此，你此时的新的，在主线程的下一次image load请求`imageView.sd_internalSetImage`，是命中不了任何缓存的

我想解决的问题时，传入imageThumbnailPixelSize时，在不改SD的源码下，怎么才能享受内存缓存呢？

**Comment by dreampiggy** [comment]
![image](https://github.com/SDWebImage/SDWebImage/assets/6919743/dddcf586-83ae-4cd0-b47c-57d79ad1f618)

加上`.waitStoreCache` options就行了，我发回给你Demo

**Comment by dreampiggy** [comment]
唉算了，如果大量简单业务老是踩这个坑，我可以在之后的SDWebImage 5.19.0改变这个behavior

默认是"waitStoreCache"，即同步等待的，可以手动改成async的

**Comment by dreampiggy** [comment]
[SDWebImageTest.zip](https://github.com/SDWebImage/SDWebImage/files/13857246/SDWebImageTest.zip)

**Comment by dreampiggy** [comment]
你真实的业务场景，这种问题常见吗，即

> 有某个completionBlock的callback时刻，立即同步触发一个新的Image Loading（即sd_setImage或者SDWebImageManager.loadImage）？

这种情况下你现在的版本，都需要加上`.waitStoreCache`，不然completionBlock的时刻，磁盘还没写完的

**Comment by dreampiggy** [comment]
其实从目前的这个`waitStoreCache`语义感觉也是一个问题，它应该只有disk 写是async的，mmeory 写应该是sync的。但是看起来不是

你的这种在completedBlock中立即请求thumbnail，就算没提供`waitStoreCache`，它也应该命中thumbnail的memory cache，不对吗（full size data cache是命中不了的）

**Comment by dreampiggy** [comment]
Is this old version bug ?

Try using SDWebImage 5.18.8 and use `.waitStoreCache` ?

**Comment by LeeZhe** [comment]
这个是我的新的控制器的代码demo，你可以把这个代码复制后工程跑起来，会发现，第二次加载相同的URL，cacheType还是从disk或者为none。
```
//
//  ViewController.swift
//  SDWebImageTest


import UIKit
import SDWebImage
class ViewController: UIViewController {
    lazy var button = {
       let aButton = UIButton()
        aButton.frame = CGRect(x: 0, y: 200, width: 200, height: 20)
        aButton.setTitleColor(UIColor.black, for: .normal)
        aButton.setTitle("模拟加载图片", for: .normal)
        return aButton
    }()
    
    lazy var imageView = {
        let imageView = UIImageView(frame: CGRect(x: 0, y: 0, width: self.view.frame.size.width, height: 600))
        return imageView
    }()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        view.addSubview(button)
        button.addTarget(self, action: #selector(loadImage), for: .touchUpInside)
        view.addSubview(imageView)
        SDImageCache.shared.config.shouldUseWeakMemoryCache = true
    }
    
    @objc func loadImage() {
        for _ in 0..<100{
            DispatchQueue.main.async {
                self.imageView.sd_internalSetImage(with: URL(string: "https://img01.sc115.com/uploads3/sc/jpgs/2007/apic26572_sc115.com.jpg"), placeholderImage: nil, options: [.waitStoreCache], context: [SDWebImageContextOption.imageThumbnailPixelSize: NSValue(cgSize: CGSizeMake(600, 900))], setImageBlock: { image, data, cacheType, imageUrl in
                    debugPrint("cacheType is \(cacheType.desc)")
                    self.imageView.image = image
                }, progress: nil)
            }
        }
        
    }
    
    
    override func viewWillLayoutSubviews() {
        super.viewWillLayoutSubviews()
        button.center = CGPointMake(view.center.x, imageView.frame.size.height + imageView.frame.origin.y + 150)
        imageView.backgroundColor = UIColor.red
        imageView.frame.origin = CGPointMake(0, 100)
    }
}

extension SDImageCacheType {
    var  desc: String {
        switch self {
        case .all:
            return "all"
        case .disk:
            return "disk"
        case .memory:
            return "memory"
        case .none:
            return ".none"
        @unknown default:
            return "unknown"
        }
    }
}
```
而在我实际的工程中，这种场景会出现每次都是从disk，或者从network里加载。

![image](https://github.com/SDWebImage/SDWebImage/assets/11971566/22662c43-be8d-4465-b8c5-eea181ce9081)

**Comment by dreampiggy** [comment]
你这个Demo的写法，不命中memory cache不是完全符合预期的吗？

memory cache只在"sd_set"（以后用Image Request指代）函数返回前，也就是进入function call的瞬间，检查一次。

你在“同一时刻”提交了100个Image Request，它同时检查因为此刻没有任何内存的缓存，内存缓存是空的

如果是第二次App运行，则可以从disk加载
如果是第一次App运行，必然全部是network

**Comment by LeeZhe** [comment]
不是的，第一次加载从network可以理解，但是第二次刷新同一个URL的时候，我期望是从memory。

**Comment by dreampiggy** [comment]
第二次它是从disk data拿到的NSData（比如是1000x1000)

然后重新decode成一个thumbnail（比如是100x100）

然后把这个disk data（1000x1000）重新覆盖原始cache key（可以忽略，这个啥都不影响）
然后把这个thumbnail image覆盖到thumbnail key

**Comment by LeeZhe** [comment]
我把SD的源码 

```
                if (thumbnailSize.width > 0 && thumbnailSize.height > 0) {
                    // Query full size cache key which generate a thumbnail, should not write back to full size memory cache
                    shouldCacheToMomery = YES;
                }

```
时，就能符合我的预期。但是不知道这个有没有其它的影响。

**Comment by LeeZhe** [comment]
> 第二次请求它会先检查thumbnail key对应的memory cache是否存在 这一步如果前一次Image Request结束，那么会一定加载成功的

但是我实际的场景是不成功，加载的是network或者disk。所以就产生了刷新页面加载同一个URL时出现了闪烁的问题。

**Comment by dreampiggy** [comment]
同一个ImageView的不同Image Request，会cancel 前者，这样写法不太好，或者说没意义

**Comment by dreampiggy** [comment]
> "cacheType is .none"
这些输出是因为前99个Image Request被你cancel了（during disk cache query）

**Comment by dreampiggy** [comment]
发现这里有一个逻辑问题

**Comment by LeeZhe** [comment]
> > "cacheType is .none"
> 
> 这些输出是因为前99个Image Request被你cancel了（during disk cache query）

是的，这个我之前模拟的时候没有考虑的，那么我就每次点击的时候就加载一次请求的话，怎么能让第二次点击是从memory里检索到呢？
```
@objc func loadImage() {
        self.imageView.sd_internalSetImage(with: URL(string: "https://img01.sc115.com/uploads3/sc/jpgs/2007/apic26572_sc115.com.jpg"), placeholderImage: nil, options: [.waitStoreCache], context: [SDWebImageContextOption.imageThumbnailPixelSize: NSValue(cgSize: CGSizeMake(600, 900))], setImageBlock: { image, data, cacheType, imageUrl in
            debugPrint("cacheType is \(cacheType.desc)")
            self.imageView.image = image
        }, progress: nil)
    }
```

**Comment by dreampiggy** [comment]
![image](https://github.com/SDWebImage/SDWebImage/assets/6919743/81e40f84-32dd-490c-87f8-06d789e785b1)

如果从Disk Cache拿到了Thumbnail Image（就是你的第100个请求）的情况下

这里不会触发额外的逻辑，直接callback回去了。所以，“不会把这个Thumbnail Image重新写回到Memory Cache”

**Comment by dreampiggy** [comment]
> 怎么能让第二次点击是从memory里检索到呢？

看起来目前没有这个策略。因为一旦从Disk Cache读到Thumbnail UIImage后，就直接返回给你了。“没有再次把这个Thumbnailk UIImage写到Memory Cache，对应Thumbnail Key”的逻辑

**Comment by dreampiggy** [comment]
是曾经有这样的逻辑吗，总感觉确实有道理。这会导致一旦你的Thumbnail从Disk读取，以后只能从Disk 读取

**Comment by LeeZhe** [comment]
> > 怎么能让第二次点击是从memory里检索到呢？
> 
> 看起来目前没有这个策略。因为一旦从Disk Cache读到Thumbnail UIImage后，就直接返回给你了。“没有再次把这个Thumbnailk UIImage写到Memory Cache，对应Thumbnail Key”的逻辑

对的，终于说通了。在我们的业务场景里，会点击商品下面的小图，点击小图之后会出现刷新页面，此时我在开启Thumbnailk时就不享受内存了，diskImage读取是耗时的，此时用户就能明显感觉到闪烁

**Comment by dreampiggy** [comment]
这个问题的逻辑和SDWebImageManager无关，你上来贴代码把问题说远了。实际上和Manager里面的什么storeImage的逻辑一点关系没有

**Comment by dreampiggy** [comment]
SDImageCache曾经有一个逻辑是：

如果query，从disk cache得到了一个UIImage，我们把它sync 回memory cache

**Comment by LeeZhe** [comment]
> 这个问题的逻辑和SDWebImageManager无关，你上来贴代码把问题说远了。实际上和Manager里面的什么`callStoreXXX`，也就是下载后的逻辑一点关系没有

我是调试之后，发现代码里的逻辑注释后，就会塞回memoryCache，此时就能享受内存缓存，所以来问你，在没有改源码的情况下，如何能享受内存缓存。

**Comment by LeeZhe** [comment]
> SDImageCache曾经有一个逻辑是：
> 
> 如果query，从disk cache得到了一个UIImage，我们把它sync 回memory cache
> 
> 这个逻辑我印象改过，需要找到当时的背景，避免regression的同时修复

我看你源码里是有注释备注的，但是我不太看得懂，不懂背景。所以我应该怎么处理才能传入thumbnail的context时，享受内存缓存呢？

**Comment by dreampiggy** [comment]
好像知道问题了。

SDImageCache是不认识thumbnail的，只**以Key**区分。这个设计背景SDWebImageManager完全不同

因此这里有一个悖论：

1. Manager先请求Cache，要求拿到 `thumbnail key` （context也不变）的image和data，此时没命中任何的缓存
2. Manager再请求Cache，要求拿到`full key`（context也不变）的image和data，此时拿到了data，以及thumbnail image

然后就没其他操作了

**Comment by dreampiggy** [comment]
当时发生的问题是：

在2. 流程中，Cache眼里的key是`full key`，但是拿到的UIImage是thumbnail image

所以和4的流程不match

**Comment by dreampiggy** [comment]
为了快速解决问题，暴力hack了一下检查thumbnail的逻辑，不让他store错误的cache到memory中，但是现在看这个问题解决的不彻底。会导致你说的disk cache永远命中的问题（而丢掉了同步到memory cache的逻辑）

**Comment by LeeZhe** [comment]
如果我把

> 但是拿到的UIImage是thumbnail image

你这里传进来的key不是thumbnail 的cache key么? 那么理论上应该拿到的diskCache也是thumbnail的， 那么扔回内存不是应该是正常的逻辑么？但是现在是不放回memoryCache。

**Comment by dreampiggy** [comment]
@LeeZhe 看邮件

**Comment by dreampiggy** [comment]
针对5.18.x版本打算先快速解决问题

针对5.19，打算彻底引入新的API，即Cache不应该自动“在Query key1得到disk image1，同步这个image1到memory cache”

而是调用方，也就是SDWebImageManager，自己把控触发一个从disk往memory同步的流程

**Comment by LeeZhe** [comment]
我这边，直接在SD的源码里把image 同步到memory cache，是能解决我的问题，此时会有别影响吗？

**Comment by dreampiggy** [comment]
@LeeZhe 有别的影响，你会导致cache写错（即用thumbnail的cache key覆盖了原始的full size的memory cache），所以我意思你先反馈问题，不要在不清楚上下文的情况下改

https://github.com/SDWebImage/SDWebImage/pull/3657 测试一下这个分支

**Comment by dreampiggy** [comment]
@LeeZhe Any update ?

我测试Demo可正常运行，包括3种场景，多次load后都可以最终收敛，到命中memory cache：
1. 都没有任何缓存（第一次启动）
2. 磁盘缓存有full data
3. 磁盘缓存有thumbnail data，但是没有内存缓存thumbnail image

**Comment by LeeZhe** [comment]
@dreampiggy 我试了你给分支， 是符合我的预期的。使用了https://github.com/SDWebImage/SDWebImage/pull/3657  分支后我这边上层需要调整。
-  需要适配在setImageBlock中判断placeHolder与之前判断不一致的问题。

**Comment by dreampiggy** [comment]
发布5.18.9版本
